pkg/controllers/internalmembercluster/v1beta1/member_controller.go

```diff
@@ -156,20 +272,183 @@ func (r *Reconciler) stopAgents(ctx context.Context, imc *clusterv1beta1.Interna
 
 // updateHealth collects and updates member cluster resource stats and set ConditionTypeInternalMemberClusterHealth.
 func (r *Reconciler) updateHealth(ctx context.Context, imc *clusterv1beta1.InternalMemberCluster) error {
-	klog.V(2).InfoS("updateHealth", "InternalMemberCluster", klog.KObj(imc))
-
-	if err := r.updateResourceStats(ctx, imc); err != nil {
-		r.markInternalMemberClusterUnhealthy(imc, fmt.Errorf("failed to update resource stats %s: %w", klog.KObj(imc), err))
+	klog.V(2).InfoS("Updating health status", "InternalMemberCluster", klog.KObj(imc))
+
+	probeRes := r.rawMemberClientSet.Discovery().RESTClient().Get().AbsPath(healthProbePath).Do(ctx)
+	var statusCode int
+	if probeRes.StatusCode(&statusCode); statusCode != 200 {
+		err := fmt.Errorf("health probe failed with status code %d", statusCode)
+		klog.ErrorS(err, "Failed to probe health", "InternalMemberCluster", klog.KObj(imc))
+		r.markInternalMemberClusterUnhealthy(imc, err)
 		return err
 	}
 
+	klog.V(2).InfoS("Health probe succeeded", "InternalMemberCluster", klog.KObj(imc))
 	r.markInternalMemberClusterHealthy(imc)
 	return nil
 }
 
+// connectToPropertyProvider connects to the property provider to collect the latest cluster properties.
+func (r *Reconciler) connectToPropertyProvider(ctx context.Context, imc *clusterv1beta1.InternalMemberCluster) error {
+	r.propertyProviderCfg.startPropertyProviderOnce.Do(func() {
+		if r.propertyProviderCfg.propertyProvider == nil {
+			// No property provider is set up; fall back to the built-in default behavior.
+			return
+		}
+
+		childCtx, childCancelFunc := context.WithDeadline(ctx, time.Now().Add(propertyProviderStartupDeadline))
+		// Cancel the context any way to avoid leaks.
+		defer childCancelFunc()
+		startedCh := make(chan error)
+
+		// Attempt to start the property provider.
+		go func() {
+			defer close(startedCh)
+			if err := r.propertyProviderCfg.propertyProvider.Start(r.globalCtx, r.propertyProviderCfg.memberConfig); err != nil {
+				klog.ErrorS(err, "Failed to start property provider", "InternalMemberCluster", klog.KObj(imc))
+				startedCh <- err
+			}
+		}()
+
+		// Wait for the property provider to start; if it doesn't start within the deadline,
+		// fall back to the built-in default behavior.
+		select {
+		case <-childCtx.Done():
```

Comment:
This is basically spawn a goroutine and wait for it to finish inline. I don't understand why not just call the function inline with the childCtx?

Reply:
This is to add a deadline to the Start() method, which is calling external implementation with no guarantees, i.e., it might block and not respect the cancellation of `childCtx`, hence the select call.